Hw2

prefabsprout opened the pull request.

prefabsprout committed:
Add test_exercise_1.py for HW2

prefabsprout committed:
Put driver initialization into conftest.py

prefabsprout committed:
Add test_exercise_2.py

zhidelev commented on test/HW2/conftest.py:
```diff
@@ -0,0 +1,16 @@
+def driver():
+    driver = webdriver.Chrome()
+    driver.implicitly_wait(time_to_wait=5)
+    return driver
```
Лучше делать `yield driver`, а после его закрывать, а не делать `close()` в тесте.

zhidelev commented on test/HW2/locators/different_elements_locators.py:
```diff
@@ -0,0 +1,9 @@
+from selenium.webdriver.common.by import By
+
+
+class DifferentElementsLocators:
```
Лучше разделить локаторы в классы Checkbox, Radiobutton, Log, а не объединятть их в класс с именем DifferentElementsLocators (слишком общее имя).

zhidelev commented on test/HW2/test_exercise_1.py:
```diff
@@ -0,0 +1,63 @@
+
+def test_exercise_1(user_credentials, driver):
+    # 1.Open browser and go to page
+    driver.get(url="https://jdi-testing.github.io/jdi-light/index.html")
```
Вынести URL в констанут и использовать через импорт.

zhidelev commented on test/HW2/test_exercise_1.py:
```diff
@@ -0,0 +1,63 @@
+    driver.find_element(*JdiTestingLocators.PROFILE_MENU).click()
+    driver.find_element(*JdiTestingLocators.USER_NAME_FORM).send_keys(user_credentials["username"])
+    driver.find_element(*JdiTestingLocators.USER_PASSWORD_FORM).send_keys(user_credentials["password"])
+    driver.find_element(*JdiTestingLocators.LOGIN_BUTTON).click()
```
Думаю, что этот код можно вынести в модуль `functions.py` например и объединить в функцию `login()` и использовать в тесте.

prefabsprout replied:
Я предполагал вынести это в отдельный метод на третьей домашке. Там как раз надо реализовать page object. А тут я подумал, что домашка требует более "влобную" реализацию на чистом селениуме

zhidelev commented on test/HW2/test_exercise_1.py:
```diff
@@ -0,0 +1,63 @@
+    driver.find_element(*JdiTestingLocators.USER_PASSWORD_FORM).send_keys(user_credentials["password"])
+    driver.find_element(*JdiTestingLocators.LOGIN_BUTTON).click()
+
+    # 4.Assert Username is loggined
```
Typo loggined -> logged

zhidelev commented on test/HW2/test_exercise_1.py:
```diff
@@ -0,0 +1,63 @@
+
+    # 8.Assert that there is the iframe with “Frame Button” exist
+    iframes_with_frame_button = driver.find_elements_by_id("frame")
+    assert iframes_with_frame_button != []
```
Возможно лучше переписать:
```
assert driver.find_elements_by_id("frame")
```

prefabsprout replied:
Я подумал, что поместить это в переменную лучше из соображений читаемости. Условие домашки потребовало найти фреймы с frame button. Из описания локатора сложно сказать, что это именно локаторы с этой кнопкой. Но это они

prefabsprout replied:
Но избавиться от `!= []` точно надо, согласен, это косяк. Избавлюсь, как только получу комментарий насчёт моей мысли оставить список из элементов в переменной

zhidelev replied:
Я не верен, что flake8 не будет ругаться на этот код, если переменная больше нигде не используется. Попробуй запустить flake8 на этот код, если не ругается, то оставляй испльзование переменной 👍

prefabsprout replied:
flake8 ругается только на длину строк:
```
>flake8 test
test\test_different_elements_page_functionality.py:23:80: E501 line too long (86 > 79 characters)
test\test_different_elements_page_functionality.py:24:80: E501 line too long (90 > 79 characters)
test\test_different_elements_page_functionality.py:28:80: E501 line too long (97 > 79 characters)
test\test_different_elements_page_functionality.py:37:80: E501 line too long (83 > 79 characters)
test\test_different_elements_page_functionality.py:41:80: E501 line too long (84 > 79 characters)
test\test_different_elements_page_functionality.py:46:80: E501 line too long (81 > 79 characters)
test\test_different_elements_page_functionality.py:49:80: E501 line too long (86 > 79 characters)
test\test_different_elements_page_functionality.py:52:80: E501 line too long (83 > 79 characters)
test\test_main_page_layout.py:17:80: E501 line too long (86 > 79 characters)
test\test_main_page_layout.py:18:80: E501 line too long (90 > 79 characters)
test\test_main_page_layout.py:22:80: E501 line too long (97 > 79 characters)
test\test_main_page_layout.py:24:80: E501 line too long (100 > 79 characters)
test\test_main_page_layout.py:26:80: E501 line too long (83 > 79 characters)
test\test_main_page_layout.py:27:80: E501 line too long (82 > 79 characters)
test\test_main_page_layout.py:28:80: E501 line too long (87 > 79 characters)
test\test_main_page_layout.py:36:80: E501 line too long (93 > 79 characters)
test\test_main_page_layout.py:37:80: E501 line too long (99 > 79 characters)
test\test_main_page_layout.py:38:80: E501 line too long (99 > 79 characters)
test\test_main_page_layout.py:39:80: E501 line too long (82 > 79 characters)
test\test_main_page_layout.py:40:80: E501 line too long (90 > 79 characters)
test\test_main_page_layout.py:41:80: E501 line too long (80 > 79 characters)
test\test_main_page_layout.py:42:80: E501 line too long (86 > 79 characters)
test\test_main_page_layout.py:43:80: E501 line too long (92 > 79 characters)
test\test_main_page_layout.py:44:80: E501 line too long (92 > 79 characters)
test\test_main_page_layout.py:45:80: E501 line too long (94 > 79 characters)
test\test_main_page_layout.py:46:80: E501 line too long (83 > 79 characters)
test\test_main_page_layout.py:52:80: E501 line too long (81 > 79 characters)
test\test_main_page_layout.py:59:80: E501 line too long (98 > 79 characters)
test\test_main_page_layout.py:61:80: E501 line too long (91 > 79 characters)
test\test_main_page_layout.py:62:80: E501 line too long (81 > 79 characters)
test\test_main_page_layout.py:63:80: E501 line too long (95 > 79 characters)
test\test_main_page_layout.py:64:80: E501 line too long (95 > 79 characters)
test\Locators\buttons.py:8:80: E501 line too long (85 > 79 characters)
test\Locators\buttons.py:9:80: E501 line too long (86 > 79 characters)
test\Locators\buttons.py:10:80: E501 line too long (83 > 79 characters)
test\Locators\buttons.py:11:80: E501 line too long (91 > 79 characters)
test\Locators\buttons.py:12:80: E501 line too long (86 > 79 characters)
test\Locators\buttons.py:13:80: E501 line too long (92 > 79 characters)
test\Locators\buttons.py:14:80: E501 line too long (93 > 79 characters)
test\Locators\buttons.py:15:80: E501 line too long (100 > 79 characters)
test\Locators\buttons.py:17:80: E501 line too long (89 > 79 characters)
test\Locators\checkboxes.py:5:80: E501 line too long (103 > 79 characters)
test\Locators\checkboxes.py:6:80: E501 line too long (102 > 79 characters)
test\Locators\dropdowns.py:5:80: E501 line too long (86 > 79 characters)
test\Locators\radiobuttons.py:5:80: E501 line too long (106 > 79 characters)
test\Locators\text_sections.py:5:80: E501 line too long (82 > 79 characters)
test\Locators\text_sections.py:6:80: E501 line too long (80 > 79 characters)
```
Но на мой взгляд пайчармовский линтер (он ругается на > 120 символов) более актуальный в контексте современных мониторов.

zhidelev replied:
Да, `E501` не критичен.

zhidelev commented on test/HW2/test_exercise_1.py:
```diff
@@ -0,0 +1,63 @@
+from locators.jdi_testing_locators import JdiTestingLocators
+
+
+def test_exercise_1(user_credentials, driver):
```
Назвать функцию более понятно согласно тому, что в ней тестируется.

zhidelev commented on test/HW2/test_exercise_2.py:
```diff
@@ -0,0 +1,50 @@
+from locators.different_elements_locators import DifferentElementsLocators
+
+
+def test_exercise_2(user_credentials, driver):
```
Назвать функцию более понятно согласно тому, что в ней тестируется.

prefabsprout committed:
Rearrange project structure

prefabsprout commented:
И да, пояснение по поводу последнего коммита (который Rearrange project structure). Я решил так сделать, поскольку остальные домашние работы подразумевают рефакторинг уже написанной "наивной" реализации. Исходя из этого и из того, что у нас ветки, я не вижу смысла нагромождать репозиторий папками - ведь домашки будут рассортированы по веткам.

prefabsprout committed:
Fix driver initializing and hide driver.quit() inside fixture

prefabsprout committed:
Group locators by type and create class for each locator type

prefabsprout committed:
Put URL in constants.py

prefabsprout committed:
Fix typo and rename tests

prefabsprout committed:
Fix different_elements_page_locators.py name to checkboxes.py

And remove redundant import

prefabsprout committed:
Fix test_different_element_page_functionality name to test_different_elements_page_functionality

prefabsprout committed:
Rename test_exercise_<number> files

zhidelev commented on test/Locators/buttons.py:
```diff
@@ -0,0 +1,17 @@
+from selenium.webdriver.common.by import By
+
+
+class Buttons:
+    PROFILE_MENU_BUTTON = (By.CLASS_NAME, "uui-profile-menu")
+    LOGIN_BUTTON = (By.ID, "login-button")
+    HOME_BUTTON = (By.CSS_SELECTOR, ".uui-header .nav > li:nth-child(1) > a")
+    CONTACT_FORM_BUTTON = (By.CSS_SELECTOR, ".uui-header .nav > li:nth-child(2) > a")
+    METALS_COLORS_BUTTON = (By.CSS_SELECTOR, ".uui-header .nav > li:nth-child(4) > a")
+    SIDEBAR_HOME_BUTTON = (By.CSS_SELECTOR, ".sidebar-menu > li:nth-child(1) span")
+    SIDEBAR_CONTACT_FORM_BUTTON = (By.CSS_SELECTOR, ".sidebar-menu > li:nth-child(2) span")
+    SIDEBAR_SERVICE_BUTTON = (By.CSS_SELECTOR, ".sidebar-menu > li:nth-child(3) span")
+    SIDEBAR_METALS_COLORS_BUTTON = (By.CSS_SELECTOR, ".sidebar-menu > li:nth-child(4) span")
+    SIDEBAR_ELEMENTS_PACKS_BUTTON = (By.CSS_SELECTOR, ".sidebar-menu > li:nth-child(5) span")
+    DIFFERENT_ELEMENTS_BUTTON = (By.CSS_SELECTOR, ".sidebar-menu > li:nth-child(3) li:nth-child(8)")
+    FRAME_BUTTON = (By.ID, "frame-button")
+    SERVICE_DROPDOWN_BUTTON = (By.CSS_SELECTOR, ".uui-header .nav > li:nth-child(3) > a")
```
Может стоит их ещё разделить на кнопки из `Header` и `Sidebar`?

zhidelev commented on test/test_main_page_layout.py:
```diff
@@ -0,0 +1,64 @@
+    assert driver.find_element(*Buttons.METALS_COLORS_BUTTON).text == "METALS & COLORS"
+
+    # 6.Assert that there are 4 images on the Index Page and they are displayed
+    assert driver.find_element(*Icons.ICON_PRACTISE).is_displayed() is True
+    assert driver.find_element(*Icons.ICON_CUSTOM).is_displayed() is True
+    assert driver.find_element(*Icons.ICON_BASE).is_displayed() is True
+    assert driver.find_element(*Icons.ICON_MULTI).is_displayed() is True
```
Необязательно делать `assert ... is True`, можно просто `assert ...`

zhidelev reviewed (changes requested):
Поправить пару комментариев 🙂

prefabsprout committed:
Divide Buttons class on SidebarButtons and HeaderButtons

prefabsprout committed:
Get rid of not PEP8-ble boolean asserts

prefabsprout committed:
Add frame button to separate class

Because frame button is not header\footer button

prefabsprout committed:
Update opening of Different elements page

Because I screwed up a little bit - according to homework we need to open that page via header section, not via sidebar
